### DISCUSSION TITLE
[Important] Improve the user experience on the public instances

### DISCUSSION THREAD

**Discussion by unixfox** [searxng/searxng#2802] [Ideas]
# Main problem

It has been reported multiple times that users lately have a hard time using SearXNG because it doesn't work for them.

Proofs:
- https://github.com/searxng/searxng/issues/2797
- https://old.reddit.com/r/Searx/comments/16j3oq5/all_searxng_instances_return_no_search_results/
- https://old.reddit.com/r/selfhosted/comments/16emfv0/4get_a_proxy_search_engine_that_doesnt_suck/jzzb70r/
- Probably more

## 1: average user seeing the main error message

If I try to be like an average user and I see this message:
![image](https://github.com/searxng/searx-instances/assets/4016501/7895454e-8a70-4efc-9d5b-4501ddad7f56)

What am I supposed to do? There is no clear path that explain what I should do.

We should bring back the old error message (from oscar) that at least explain to use searx.space for another instance:

![image](https://github.com/searxng/searx-instances/assets/4016501/0ad136ad-c285-467d-813f-9cb6ae0d6d69)

Even better, with one click have a way to easily switch to another random SearXNG instance for example, Invidious has this support:

![image](https://github.com/searxng/searx-instances/assets/4016501/44ca8d55-1b8b-4601-8c66-fe20d6ab0145)

## 2: we are hiding the errors too deep down

And we should not hide too much the errors message, like on this page I only get results from wiki because all the other engines do not work. What am I supposed to do if I see there are no meaningful results except the ones from wikipedia? That's a real question that someone asked on reddit on the second link above.

<img src="https://github.com/searxng/searx-instances/assets/4016501/671cd3a5-eb4a-43e4-a29d-aede077754ec" width="800" />

Or we categorize the wiki engines as low priority engines, and we display the error message anyway if we find that no other engines are working. Idea: The results from the wiki engines will **ONLY** work when other engines are returning results.

Personal opinion/rant: These wiki engines have long been an issue for me, on searx.space they falsely advertise that the instance work because these wiki engines always work. Look on searx.space the amount of instances that only return wiki results and are still categorized as "working instance" that's insane. And I don't get why they exist, existing search engines like google or duckduckgo already return results from wikipedia and much more smart in terms of keywords/context than searxng.

And why do we need to have the wiki results duplicated, we have it on the right side and also in the results list?

<img src="https://github.com/searxng/searx-instances/assets/4016501/84ba81c2-e287-4c63-89a9-167d47b9b6ce" width="800" />

# Solution proposal

### 2) (Unsure, to discuss) Enable by default at least one engine that is safe to never block searxng. It's better to at least get results than nothing.

Or have backup engines that only activate when all the default engines are suspended or fail to return results?

### 3) Have a dedicated setting in SearXNG just for public instances

This setting will enable the advanced limiter, maybe allow us to set more harsh limits in the limiter and be able to add more things in the future that we know will only affect public instances.

### 4) Explain to the user that he can also host SearXNG locally.

Many people do not know that they can use SearXNG locally. They get frustrated because SearXNG do not work, but they do not know that they can use the software locally, which probably will work much better.

On this side, whoogle is leading on the easiest ways to use their software. I won't say we should reach their level, but we should at least have an easy way to deploy searxng **locally not on a remote server**, **searxng-docker is not a candidate for a local install**.

### 5) (Unsure, to discuss) Recategorize and rethink the implementation of the wiki engines into searxng

Like stated above, they probably take too much "priority" into SearXNG. They should IMO only exist on the right side of SearXNG only when the keyword is matched and if we need more wiki sources then we let the search engines dealing with that. They have better algorithms to understand which results to show than us.

IMO in the General category, for the results list (not the right side) there should exist by default only search engines that have a smart/advanced algorithm for search queries. Wiki engines do not have this kind of algorithm into their search API.

Or we create a new separate category called Wiki?

The conclusion is that in the general category, that's where most of the SearXNG users will ever use. They want something that work **and** work **LIKE** a search engine, so when they see many irrelevant results from wiki engines, they are most likely lost.

The general category should be the category where we think and discuss together when adding or removing engines, this is THE place for any beginner in SearXNG.

### 6) Only allow instances with the limiter activated (or an alternative limiter) on https://searx.space

There are many public instances that do not give results simply because the limiter is not activated, and they got abused by bot all day. We should maybe think into only allow instances that have the builtin limiter activated or a proof that they have an alternative limiter in place.

For easily finding if an instance have the limiter, we could have this setting available from a special endpoint, like through a similar page like /stats.

---------------

### This should be our number one priority, if our main core feature do not work (search for results) then what's the point in improving the (bonus) user experience with additional features.

**Comment by unixfox** [comment]
ping @dalf @return42 @Bnyro @mrpaulblack @tiekoetter @ononoki1

**Comment by return42** [comment]
My unsorted remarks

I can agree with the most written .. but I have a different view on some points

> Bring back the old error message and improve it so that the user know what to do.

the old messages are still there, they are just not displayed if you don't click on the box... and that's good, because these messages only cause uncertainty for an ordinary user.

> (Unsure, to discuss) Enable by default at least one engine that is safe to never block searxng. It's better to at least get results than nothing.

This is not possible, if the user selects one engine (or the maintainer disables all engines in the "general category" except google) and this engine blocks this is the result:

![image](https://github.com/searxng/searxng/assets/554536/067d863c-a29c-45f9-8d0c-d90dd79a4c24)




if an instance is poorly maintained and the settings are set unfavorably, then you just don't have any results, that's the way it is ... Instances must be maintained and inexperienced users should not play around too much in the settings.

**Reply by unixfox** [reply]
> My unsorted remarks
> 
> I can agree with the most written .. but I have a different view on some points
> 
> > Bring back the old error message and improve it so that the user know what to do.
> 
> the old messages are still there, they are just not displayed if you don't click on the box... and that's good, because these messages only cause uncertainty for an ordinary user.

What do you think is good between : 

- The user being frustrated because the searxng instance doesn't return any meaningful results (no engines are working, only wiki engines return results).
- The user having good results but a red box showing some errors.

There is a tradeoff to do here, I don't think I have ever received any report of users being afraid of the error messages? Did you get these kinds of reports? I'm talking about real user experience here, not guessing what the average user would think of it.

From my experience in 5 years of working with both SearX and SearXNG I have never seen anyone being afraid of these errors, in fact this has boosted our amount of reports of broken engines. I know that's something you did not like, the situation has improved, but that's part of maintaining a public project! I prefer to have reports of broken engines than angry people that say "nothing works!".

My idea: I propose to display the old error message improved when there is only 1 engine (or none) out of 3 (or more) that returns the results. Or a percentage, if only 80% of the engine do not work.

We can improve the error message like this:

```
Engines cannot retrieve results:
engine1 ( Suspended: access denied )
engine2 ( Suspended: access denied )
engine3 ( HTTP error )

Only 1 engine is working. If you are an user, please consider switching instance: https://searx.space
If you are the maintainer please consider disabling these engines or changing the IP address of your server.

```

> > (Unsure, to discuss) Enable by default at least one engine that is safe to never block searxng. It's better to at least get results than nothing.
> 
> This is not possible, if the user selects one engine (or the maintainer disables all engines in the "general category" except google) and this engine blocks this is the result:

Maybe forget about backup engines but enabling by default an engine that will most likely not block searxng is a good idea.

> if an instance is poorly maintained and the settings are set unfavorably, then you just don't have any results, that's the way it is ... Instances must be maintained and inexperienced users should not play around too much in the settings.

When did we stop caring about listing only working instance on https://searx.space that have a great user experience? This is not the mindset we have always kept during these years.

**Reply by return42** [reply]
> What do you think is good between :

May be the best compromise is in between  --> I think we can improve this message .. 

![image](https://github.com/searxng/searx-instances/assets/4016501/7895454e-8a70-4efc-9d5b-4501ddad7f56)

.. by adding a hint like *"If you often see this message we recommend to use a other instance from searx.space"*

> There is a tradeoff to do here, I don't think I have ever received any report of users being afraid of the error messages? Did you get these kinds of reports? 

Yes, real users / I mean "really" real users without any knowledge of IT .. 

> I know that's something you did not like

Yes & No, I will explain what I have in mind .. I have seen three kind of user groups:

1. The group of users who move anxiously and inexperienced through the internet that is unsettled by those error messages // Mostly these are older people who have heard bad things about the internet ... about scams and other bad things. I myself am a little older and believe me in my circle of friends there are some of them / which I recommend SearXNG --> imagine you had no idea about the internet and every second google search would always come up with red error messages like this.

2. The group of user that do not read hints but press every button it can find, in the hope there is a solution or they can help in any manner.  This was the user group that did not fill out bug reports on github / I have seen bugs from users that created a github account just to leave a blank form in our issue tracker .. they have never used there github accounts any more and they did not give feedback to their issues ..  we can't blame them, with the red error message and the github button we are directing uninitiated users to a developer platform (github) that they don't know and don't understand.

3. The experienced group of users, who are able to expand the field with the red error messages and then know what to do ... since we have to expand the field we have higher quality error messages and the noise caused by group 2 is no longer present.

> Maybe forget about backup engines but enabling by default an engine that will most likely not block searxng is a good idea.

I drive a little bit different concept on my instance and this is also what we are reflecting with our default setup / I refer explicitly to the category "general" but actually the concept applies to all categories :

- Activate all major search engines that are able to return a result in less than three seconds.
- Ignore temporary error messages from the engines / it happens from time to time that a search engine does not respond in the given time or there are any other problems with it.

The last point I will explain by a simple example:

![image](https://github.com/searxng/searxng/assets/554536/282adfd4-2ecd-4e88-a10b-bb07648a0de4)

As we can see my qwant engine has temporarily issue .. but does it matter? --> No: The percentage is below the measurement limit / zero percent --> we can ignore

Since we have bing, brave, ddg, google, mojeek, qwant, startpage and more active in the "general" category the user will always get results .. lets fail engines temporary, it doesn't matter if we make sure that all search engines work as well as possible.

The fact that errors are temporary is one more reason not to push them into the user's hit list in red error messages.

**Reply by unixfox** [reply]
> May be the best compromise is in between --> I think we can improve this message ..
> 
> ![image](https://user-images.githubusercontent.com/4016501/268496116-7895454e-8a70-4efc-9d5b-4501ddad7f56.png)
> 
> .. by adding a hint like _"If you often see this message we recommend to use a other instance from searx.space"_

https://github.com/searxng/searxng/pull/2808 :)

> > There is a tradeoff to do here, I don't think I have ever received any report of users being afraid of the error messages? Did you get these kinds of reports?
> 
> Yes, real users / I mean "really" real users without any knowledge of IT ..
> 

Sadly, we do not have stats of the usage of our product, and we will never consider doing so.

But we should concentrate our energy, our time, our way of building this software to 95% of our user base. If we adapt or spend too much time to the small percentage of our user base (5%) we are going to burn out, receive many unhappy reports and disappoint the people that are trusting our software.

The average Joe has, I call it, or the average SearXNG user, know a little bit about how to use a software on a computer. How to troubleshoot minor issues in IT and has seen Windows errors in the past. He has probably found SearXNG through a YouTube video or a blog article.

If we hide them the solutions for him to solve his issues, how is he going to appreciate and want to use our software as a daily driver? If he gets frustrated because there is no way to solve that: "no results" error.

While I understand, this may scare people that do not know a lot about the computer and has been recommended SearXNG by someone good in IT, but this is a very minor percentage of our user base. People that are scared of errors probably know someone in IT and can always ask for this person good in IT to modify the source code of SearXNG to remove the errors.

About the user base that report just unfilled templates, you would think that these are big a chunk of our community as you keep getting these errors! But no, they are just a minority of our user base, they may be the loudest because they are annoying. But the majority of users behave normally.

While this not SearXNG community but Invidious one, which are most likely very similar, everytime Youtube would break Invidious the vast majority of people behave normally while the other one, the minority one, scream:
- https://github.com/iv-org/invidious/issues/1838
- https://github.com/iv-org/invidious/issues/3822
- https://github.com/iv-org/invidious/issues/2577

This is an issue across ALL open source projects (even popular ones like KDE), BUT it should force us to make it more difficult for people that behave normally to report the problems or **at least get a confirmation that there is indeed an issue**. There are people that behave normally and have never used Github! Usually these people will ask what's going on, on reddit because they do not understand that the engines do not work because we do not tell them nor give solutions:
- https://old.reddit.com/r/Searx/comments/16j3oq5/all_searxng_instances_return_no_search_results/
- https://old.reddit.com/r/Searx/comments/156enhw/new_to_searx_messages_from_the_search_engines/
- https://old.reddit.com/r/Searx/comments/12u584w/i_never_get_any_search_results_when_using_searx/ (not using SearXNG but you get the point)

> 
> > Maybe forget about backup engines but enabling by default an engine that will most likely not block searxng is a good idea.
> 
> I drive a little bit different concept on my instance and this is also what we are reflecting with our default setup / I refer explicitly to the category "general" but actually the concept applies to all categories :
> 
>     * Activate all major search engines that are able to return a result in less than three seconds.
> 
>     * Ignore temporary error messages from the engines / it happens from time to time that a search engine does not respond in the given time or there are any other problems with it.
> 
> 

But this is going to considerably slow down the default user experience, even more in a local installation. On your instance the response time is still high but it's ok because you have engines suspended but if we had all the engines working that would introduce so much load on SearXNG and make the loading time very long!

People are used to fast and responsive search engines, ideally we should not offer by default a user experience where the page takes more than 1.5 second to respond.

**Comment by BernieHuang2008** [comment]
I think your suggestions about randomly picking another instance is very good.

For more, we can search for an instance that can return the required results. 
For example, if ddg didn't respond, we shall choose a instance that can return ddg result and switch to it. I think it's better than just randomly switch one. (switch it automatically is also a great idea.)

I don't know if I'm clear.

**Reply by unixfox** [reply]
I think it would be difficult to provide a way to switch to an instance that have ddg working.

Mostly because you can have multiple engines enabled, and this would require to store at a certain time across all the instances which engine work and which do not. That would complexify the implementation.

I prefer to avoid an automatic switch, the user may just want to enable/disable the engine in the preferences before wanting to switch to another instance.

**Reply by BernieHuang2008** [reply]
Emmm, or we can switch to a instance with high ranking on searx.space. this will also encourage maintainer to maintain their instances.

**Reply by unixfox** [reply]
The issue is that instances are sorted by response time, so instances that have servers close to the server of searx.space will get ranked higher.

We want to spread the users around all the instances not just the top10, this wouldn't make sense to have a list that have 90 instances then.

**Reply by Bnyro** [reply]
The ranking on searx.space doesn't really have much to say on whether the instance is maintained well and working pretty good in general or not.

It's possible that an instance has a 300ms response time and is always operating, but in fact just either disabled a lot of engines or is blocked by many of them.

And usually, when you have a downtime on your instance, that's not because of yourself but rather because of issues with the hosting provider. That doesn't mean an instance is poorly maintained.

**Reply by BernieHuang2008** [reply]
But if you've redirected the user to another instance and he finds out that it is also unacceptable, what will he think?

We must make sure that the second instance is useable in order to save our users!

**Reply by unixfox** [reply]
In this case, we only redirect to instances that are marked as responding/returning results on https://searx.space. That's already implemented in the API of searx.space.

**Reply by return42** [reply]
> We must make sure that the second instance is useable in order to save our users!

IMO this comment hit the nail : we can't! 

--> All kind of ranking we can consider and implement is just a snapshot from a point-of-view and in a the moment we gather the data for a ranking .. it can be total different when a user tries it a second later from a different IP / personally I think any kind of ranking we do, give the users the wrong illusion there is one instance that is better than the other .. but probably it isn't :o

**Reply by BernieHuang2008** [reply]
Ok the quality of instances in searx.space is good.

I just thought we could do it better .... 
```
 😔
👉👈
```

**Reply by unixfox** [reply]
From my experience in the implementation of this "switch" for invidious, the users really like the functionality, to the point that some people prefer using invidious over piped just for this feature. We have seen a lot of people using it.

My point here is to give the users clear available choices, not hiding the choice behind a dropdown or showing just "nothing" then we become back to the major issue have existed in searx:
https://github.com/searx/searx/issues/2053 (example issue but there are many others). The users are not dumb, they can make choices if needed. If they leave while having choice to solve the problem, then it's up to them, but we have done our job in trying to keep this user.

I do not want to come back to a state like this, certainly not. Users hated it.

**Comment by BernieHuang2008** [comment]


**Reply by unixfox** [reply]
That's already done in searxng.

![image](https://github.com/searxng/searxng/assets/4016501/59648bbd-5fbd-483a-a9c4-8d122c0b0d33)

The point is when only the wiki engines work.

**Comment by dalf** [comment]
> 1) Bring back the old error message and improve it so that the user know what to do.

IMO, the solution is to ask the user to make a choice:
* I ignore the message (note: it will remains in written in red).
* I want to report an issue --> that's GitHub button "Submit a new issue" from the `/stats`
* this engine is broken for my usage, I want to disable it right now.

A quick UI prototype to add buttons:
![image](https://github.com/searxng/searxng/assets/1594191/0959aead-81ea-45cc-aacd-ca009686876a)

When the user clicks on it, the engine is disabled as the user can do it in the preferences.
IMO, it is important to keep the result page, so the user remains in its flow.

Technically:
* when JS enabled, the result page remains, the engine is disabled by a JS HTTP request.
* without JS, a popup can be a workaround.

**Reply by unixfox** [reply]
That's a good idea, but doesn't solve the issue when there are no more available engines that are activated. Which is an issue across all the instances right now on https://searx.space

I have said in https://github.com/searxng/searxng/discussions/2802#discussioncomment-7026518 that it would be maybe a good idea to offer fallback engines? On the same UI we could show some of the popular engines to activate.

I think these are very good ideas which will go a long way into improving the user experience.

**Reply by dalf** [reply]
So, a possible extension of https://github.com/searxng/searxng/pull/2808 is to display a list of working engine in addition to the error message. A click on the engine name sends the same query with this engine instead of the current selection.

The server knows which engines are reliable since there are reliability stats.

**Reply by unixfox** [reply]
Not a bad idea, I like the idea :+1: 

Now how to implement it properly is the real question.

**Reply by glanham-jr** [reply]
Suggestion: In submit `Open Issue`, we should also add some metric counting that someone wanted to report an issue. In the event the user goes to the Github page and decides they don't want to bother writing a report, at least the server admin will have a count of this number. With that metric, we could add an option to alert the admin via some means number of reported issues have surpassed the configured amount.

**Comment by unixfox** [comment]
> they also rely on various probes and monitoring to make sure everything works as intended. 

You're right! The only way for SearXNG there's a problem is user report! So can we have a sort of "quick report button" that allows user report this to the admin?

I'm not saying report to github issue, but to the instance admin.
We can have a count in each instance: if many users reports a same error, than we shall inform the instance admin to solve this problem(such as network problem) ASAP. 

If the admin solved it quickly, we can mark it as `well-maintained`. We can encourage admins to maintain their engine in this way.
If he don't, we can do some sort of punishment(I don't know if it's good, do or not depends on you).

This will improve the quality of each instance I think.

Without good maintaining, everything is treating symptoms and not the root cause.
( although I think this is another topic, I strongly advise you to focus on how to encourage admins to maintain their instances.)

_Originally posted by @BernieHuang2008 in https://github.com/searxng/searxng/issues/2808#issuecomment-1723151041_

**Reply by unixfox** [reply]
Most of the time, the engines issues are only solvable by the maintainers of SearXNG.

We have already discussed in the past to remove the instances that won't return any results or heavily broken, but this hasn't been decided yet. I don't want to complicate the process, we already have limited time to work on the project.

I would prefer to have a simple rule, if the maintainer of the public instance fail to maintain any results by default on his instance for one week then the instance is removed that's it. We already lower the instances that do not return any results on searx.space but it's a bit broken for the moment as the wiki engines always return some results: https://github.com/searxng/searx-space/issues/159

**Reply by BernieHuang2008** [reply]
> if the maintainer of the public instance fail to maintain any results by default on his instance for one week then the instance is removed that's it. 

Agree with you.

**Comment by glanham-jr** [comment]
> 2) (Unsure, to discuss) Enable by default at least one engine that is safe to never block searxng. It's better to at least get results than nothing.

Forcing enabling by default might not be the move. What if we have engines specifically for `fallback-only`: meaning it cannot be used in a primary request, but can only be used once < X results are returned. This is only a mitigation, and will also eventually fail as well. With this, we could also introduce steps to fallback. i.e. if fallback-engine 1 fails, then go to fallback engine 2, and so on so forth as configured.

Not sure if this is a particularly effective idea for this, but what if we allow searxng to proxy to other searxng instances? Implementation details and edge cases aside, having some backup private instances running to allow a fallback of a primary searxng node might be useful with the above logic. `This searxng node is failing? Try the next one I have setup`. Could introduce some restriction of this proxying by some API Key and preventing endless proxy loops, and possibly "proxy-only" instances which don't have a front-end exposed.

Doesn't solve the bot problem of the other thread.

**Reply by return42** [reply]
> but what if we allow searxng to proxy to other searxng instances?

I would describe what you are talking about here as "SearXNG network" .. We are currently pursuing a different concept of proxy routing for outgoing requests, which we have implemented in the so-called "SearXNG network stack".

In SearXNG you can configure Proxies for the outgoing HTTP requests.

https://github.com/searxng/searxng/blob/3535377c9aa777d95888a853d813e94fbbae5489/searx/settings.yml#L183-L186

The network stack is currently being revised:

- https://github.com/searxng/searxng/pull/2685

One of the intentions of the revision is that we can map requests such as:

> This ~searxng node~ proxy is failing? Try the next one I have setup. 

already in the network stack without the user having to select anything.

**Reply by unixfox** [reply]
I think @glanham-jr is talking about all the searxng instances becoming a decentralized network where each other can request for other instance. Like the concept of "torrent" or how the matrix protocol works.

There was a discussion in some github issues at searx repo like this one: https://github.com/searx/searx/issues/39

But ultimately I think this is a bad idea because this will mean that all the instances will be busted. It takes just a couple of bad bots to ruin the other instances as well because they are doing too many requests.

Anyway, the ideal outcome of this is to make it easier for people to switch between actual searxng instances. Invidious, another project, already offer this functionality and people love it.

**Reply by glanham-jr** [reply]
Apologies, I didn't mean to imply a complete decentralization / federation of all of the instances. I was suggesting to provide redundancy for selected (and consenting) instances. The use cases would be if I owned multiple instances and want to setup a primary -> fallback nodes, or a group of admins decide they want to link nodes together to share the burden of fallback. 

Searxng nodes would be linked via an API endpoint with an API Key, meaning unless that API key is shared, then other instances won't be able to call the endpoint. This endpoint could also detect loops and break if a previous node was met.

But yes, still doesn't solve the bot problem per-say.

---

Regardless of searxng proxying, having fallback engine(s) is definitely a minor mitigation that could help out, and shouldn't be too much effort to implement. I would consider working on this soon - trying to be more involved in OSS this year 😄 

> Anyway, the ideal outcome of this is to make it easier for people to switch between actual searxng instances. Invidious, another project, already offer this functionality and people love it.

Agreed. I also think it would be fun to use the `I'm feeling lucky` button to quickly select a random instance, in case the user doesn't want to go to https://searx.space/ and decide themselves.

**Reply by glanham-jr** [reply]
Same point was made in the other thread. Of course CAPTCHA or invite will block the bots and there is no disagreement that is the quickest and best solution. Of course redundant nodes would be an over engineer, but no work has even been started and it was made clear it wouldn't entirely solve the bot problem. 

No disagreement on my end - ideally just have CAPTCHA just be opt in and let the server admins decide. But that decision is on the maintainer - if there is a privacy respecting CAPTCHA, then I see 0 reason not to include it as an optional feature. But no need to get worked about it lol.

**Reply by BernieHuang2008** [reply]
I don't know if I've tell you before, but I have an idea.
For example, now google sends us a CAPTCHA, then the instance can send the same CAPTCHA (from google) to a mostly Bot-like client(use some .. way). If that client is a real user, he/she can solve the captcha for us, and benefit other users. They won't blame us because they, indeed, sent many requests to the instance.
If that client is a bot client, then congratulations, we've detected it via google's captcha.

**Reply by glanham-jr** [reply]
That's essentially the same as the circuit breaker logic I mentioned. Haven't dealt with bots running a private instance, but going by what lolcatw said - pretty much wouldn't do much against bots as once a searxng instance is verified, then the bots would just flood yet again. Now with the con that it only applies to a single engine and a lot of custom integration work.

**Reply by Austin-Olacsi** [reply]
What do you guys think about a "proof of compute" solution? something like this https://github.com/mCaptcha/mCaptcha
for searxng users to prove they aren't bots.

**Reply by return42** [reply]
> Just block the bots!!!!

@lolcatw FYI bot protection is what I work on --> https://docs.searxng.org/src/searx.botdetection.html (engagement is welcome)

My future plans / I'm currently working on: I collect identified bot IPs from my instance here --> https://github.com/return42/sandbox/tree/main/data/searxng  (ATM only IPv4). My intention is to import these known IP ranges  into the redis DB .. and in SearXNG incoming IPs can be lookup / verified against the IP ranges in the DB --> https://github.com/searxng/searxng/pull/2501

> Searxng hasn't been user friendly since the botting problem. Users needs to switch instances every week.

Not sure from where you have this user experience .. my instance https://darmarit.org/searx/search is public since years and well known by every bot on planet earth .. but with bot protection (aka limiter) it works flawless .. Me, my family and my friends use the instance for all search task .. and we have thousands tasks a day / SearXNG is preinstalled on any device in the home.

**Reply by unixfox** [reply]
Image or sound CAPTCHA based are like I said a no-go for me.

It's too hard to make a good CAPTCHA. I don't want to exclude groups of people.

People that could be excluded:
- doesn't know what are the images presented (cultural differences?)
- the language for the question is not a language that the user know. (Missing translations?)
- blind people. They can't see the images

---------

At the end of the day, using a public searxng instance is not much different than using a public VPN or a public proxy or using (exit nodes on) tor.

You are on the same network as other people. And there other people may not behave nicely which can result to some services being inaccessible.

That is why to me it is important to implement these features:

- provide easy switch between instances. For finding networks that are less filled with bad actors.
- allow the user to solve the engine CAPTCHA/challenge themself, thus benefitting to other users. Just like they would on a public VPN or public proxy. They are not required to do it, they have the choice. If they don't want to do that then they can use another engine.
- provide fallback engines. They are less targeted engines that may also be working.

Most user are looking for convenience, if they have buttons to switch instances or use other engines in one click they will be much more happier than having to solve CAPTCHA everyday for doing their searches regardless of the engine used.

**Reply by BernieHuang2008** [reply]
> - allow the user to solve the engine CAPTCHA/challenge themself

This is precisely what I'm saying.

By the way, is this true?
> since once you're blocked for doing too many requests on google, you're stuck in captchatown and you will get one of these on every request, regardless of how many you solved.

I think google captcha isn't like this ... well ... a few years ago.

**Reply by unixfox** [reply]
> I think google captcha isn't like this ... well ... a few years ago.

We are diverging a bit, but once you solve the google reCAPTCHA, you need to keep the cookies for the CAPTCHA solved. And these cookies are valid for a couple of hours.

Then you need to pass the CAPTCHA again.

**Reply by glanham-jr** [reply]
Is including javascript a hard no - even if the code is licensed AGPL? Even the fervorous FSF have [JS License Web Labels](https://www.gnu.org/licenses/javascript-labels.html) and another similar article [here](https://www.gnu.org/software/librejs/free-your-javascript.html), compatible with their strict [LibreJS](https://www.gnu.org/software/librejs/) plugin. If so, `mCAPTCHA` looks like it checks every box of a non image/sound based CAPTCHA.

**Reply by return42** [reply]
> Is including javascript a hard no -

Its AGPL everyone is free to do what he wants to do with the code (except license violations)

And "yes" we already have JS on client side .. but its not yet mandatory:

The thing is that we have a user base and they have legitimate expectations, e.g. because we promised them .. not to track .. etc. (?) .. AFAIK

- SearXNG has a (wide?) user base that do not activate JS.
- users that do not accept cookies 
- some might use terminal browsers (without CSS)

By example / when we activate (effective) bot detection (based on CSS) the last mentioned group can no longer use the instance ... since all public instances have to run bot blocking in order not to degrade themselves, this group can no longer participate in public instances ... if we set JS as a condition or introduce CAPTCHA or similar we will exclude further groups from public instances.

Yes, we can do anything, but we should choose the means carefully so as not to gradually exclude our user groups from use ... everyone may decide for themselves where the golden path is to be found ... in the default of SearXNG we have to find a basic configuration that is as open as possible and at the same time ensures that the instance does not degrade.

For myself, I have decided to focus more on the further development of detection and defense against bots ... but that does not exclude that there may also be other sensible options as discussed here.

**Reply by glanham-jr** [reply]
I think that was the implicit goal that needed to be expressed: maximum interoperability. That is a different goal than UX and privacy.

Best I can think of with those requirements is:
1. Have JS enabled? Use mCAPTCHA for seamless experience.
2. JS disabled? Solve a traditional CAPTCHA and post back the results.

Still hurts the experience of those choosing to disable JS, but doesn't block them. Would exclude terminal users unless you can build a savvy text only captcha that actually works that bots can't figure out easily.

Other than that, I suppose to reach your goal you'd have need to have solid server side detection with what info you have.

**Reply by return42** [reply]
> Would exclude terminal users unless you can build a savvy text only captcha that actually works that bots can't figure out easily.

Since I have been working intensively with the detection and defense of bots for some time now, I have been able to get a rough picture of the bots ... the ones that I notice and that are responsible for a degradation of the engines are usually implemented very simply ... 

What I would like to express: for a good and simple starting point, it will probably be enough to use a very simple (less robust) method to differentiate between bots and humans. We could use this method to gain experience and if we realize that we will one day be overrun by far more intelligent bots, then we can tighten the reins even further .. 

TL;DR; --> Don't try to develop a perfect method for bot detection with your first shot.

**Reply by glanham-jr** [reply]
Suppose I should start adding /s around my non serious ideas lol - a terminal captcha is definitely no where near the first choice to make.

With what you said, that will still be going along bot detection first. I don't have a public instance to provide analysis, so I can't offer much help there.

**Reply by glanham-jr** [reply]
@return42 Digging more into server side detection, I've found these two fingerprinting techniques which may be useful:

- [TLS Fingerprinting](https://lwthiker.com/networks/2022/06/17/tls-fingerprinting.html)
- [HTTP/2 Fingerprinting](https://lwthiker.com/networks/2022/06/17/http2-fingerprinting.html)

Additional report located [here on HTTP/2 fingerprinting](https://www.blackhat.com/docs/eu-17/materials/eu-17-Shuster-Passive-Fingerprinting-Of-HTTP2-Clients-wp.pdf) referenced from the blog.

Both of these are available by the author's [TS1](https://github.com/lwthiker/ts1) hash algorithm. Seems like this would be a great addition to what you currently have. But we'd definitely need to ensure that we don't accidentally block less common browsers (including terminal browsers), so we'd need to go through and test the hashes generated by all of them + collect some stats on real public servers. As per usual, the cat and mouse will catch up to this, but more basic bots would be filtered out.

**Reply by return42** [reply]
Botdetection in SearXNG is a WAF (gathering suspicious IPs) while TLS fingerprint is done on the first incoming proxy.
